src/NHibernate/Util/ExpressionsHelper.cs

```diff
@@ -15,5 +27,572 @@ public static class ExpressionsHelper
+					mappedType = null;
+					entityPersister = null;
+					component = null;
+					return false;
+				}
+
+				currentEntityPersister = convertPersister; // ((Subclass)q).Id
+			}
+
+			return TraverseMembers(
+				sessionFactory,
+				memberPaths,
+				currentEntityPersister,
+				out mappedType,
+				out entityPersister,
+				out component,
+				out memberPath);
+		}
+
+		private static bool TraverseMembers(
+			ISessionFactoryImplementor sessionFactory,
+			Stack<MemberMetadata> memberPaths,
+			IEntityPersister currentEntityPersister,
+			out IType mappedType,
+			out IEntityPersister entityPersister,
+			out IAbstractComponentType component,
+			out string memberPath)
+		{
+			// Traverse the members that were traversed by the TryGetAllMemberMetadata method in the reverse order and try to keep
+			// tracking the entity persister until all members are traversed.
+			var member = memberPaths.Pop();
+			var currentType = currentEntityPersister.EntityMetamodel.GetPropertyType(member.Path);
+			IAbstractComponentType currentComponentType = null;
+			while (memberPaths.Count > 0 && currentType != null)
+			{
+				memberPath = member.Path;
+				var convertType = member.ConvertType;
+				member = memberPaths.Pop();
+
+				switch (currentType)
+				{
+					case IAssociationType associationType:
+						ProcessAssociationType(
+							associationType,
+							sessionFactory,
+							member,
+							convertType,
+							out currentType,
+							out currentEntityPersister,
+							out currentComponentType);
+						break;
+					case IAbstractComponentType componentType:
+						currentComponentType = componentType;
+						if (currentEntityPersister == null)
+						{
+							// When persister is not available (q.OneToManyCompositeElement[0].Prop), try to get the type from the component
+							currentType = TryGetComponentPropertyType(componentType, member.Path);
+						}
+						else
+						{
+							// Concatenate the component property path in order to be able to use EntityMetamodel.GetPropertyType to retrieve the type.
+							// As GetPropertyType supports only components, do not concatenate when dealing with collection composite elements or elements.
+							// q.Component.Prop
+							member = new MemberMetadata(
+								$"{memberPath}.{member.Path}",
+								member.ConvertType,
+								member.HasIndexer);
+
+							// q.Component.Prop
+							currentType = currentEntityPersister.EntityMetamodel.GetPropertyType(member.Path);
+						}
+
+						break;
+					default:
+						// q.Prop.NotMappedProp
+						currentType = null;
+						currentEntityPersister = null;
+						currentComponentType = null;
+						break;
+				}
+			}
+			
+			// When traversed to the top of the expression, return the current tracking values
+			if (memberPaths.Count == 0)
+			{
+				memberPath = currentEntityPersister != null || currentComponentType != null ? member.Path : null;
+				mappedType = GetType(currentEntityPersister, currentType, member, sessionFactory, out _);
```

Comment:
Are you using `GetType` method in other PRs? If no, I think the last argument needs to be removed.

Reply:
No, I've removed it.